pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go

```diff
@@ -0,0 +1,302 @@
+package angulardetectorsprovider
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/plugins/config"
+	"github.com/grafana/grafana/pkg/plugins/log"
+	"github.com/grafana/grafana/pkg/plugins/manager/loader/angular/angulardetector"
+	"github.com/grafana/grafana/pkg/services/pluginsintegration/angularpatternsstore"
+)
+
+const (
+	defaultBackgroundJobInterval = time.Hour * 1
+
+	cacheRestoreTimeout = time.Second * 10
+	gcomFetchTimeout    = time.Minute * 1
+)
+
+// backgroundJob implements the dynamic angular detectors provider job that is periodically executed in the background.
+type backgroundJob interface {
+	// runBackgroundJob updates the dynamic angular detectors from a source.
+	runBackgroundJob(ctx context.Context)
+}
+
+var (
+	_ backgroundJob  = &Dynamic{}
+	_ DynamicUpdater = &Dynamic{}
+)
+
+// Dynamic is an angulardetector.DetectorsProvider that calls GCOM to get Angular detection patterns,
+// converts them to detectors and caches them for all future calls.
+// It also provides a background service that will periodically refresh the patterns from GCOM.
+type Dynamic struct {
+	log log.Logger
+
+	httpClient http.Client
+	baseURL    string
+
+	// store is the underlying angular patterns store used as a cache.
+	store angularpatternsstore.Service
+
+	// detectors contains the cached angular detectors, which are created from the remote angular patterns.
+	// Use setDetectors and ProvideDetectors to write/read this value.
+	detectors []angulardetector.AngularDetector
+
+	// mux is the mutex used to read/write the cached detectors in a concurrency-safe way.
+	mux sync.RWMutex
+
+	// initialRestore is a channel that can be used to await the initial restore to be done.
+	// It is closed when the initial restore from cache is completed.
+	// To wait for the initial restore to be done, simply read a value from this channel.
+	initialRestore chan struct{}
```

Comment:
It could be done with the `mux`, but it would require locking the `mux` before starting the [goroutine responsible for the initial restore](https://github.com/grafana/grafana/pull/70693/files/080114d63ce67f5c49c6010c9270b2ce1a383dd5#diff-a24d5d308965e3b364ed7a21123ba8917b5f644d9940230e9de564bb82e63f76R83) in `ProvideDynamic`, because there is no guarantee that the goroutine doing the initial restore runs before `ProvideDetectors` (where we await the initial restore) is called.

If `ProvideDetectors` runs first, the `mux` is unlocked and it will return `nil`, which we don't want

Reply:
> it would require locking the `mux` before starting the [goroutine responsible for the initial restore](https://github.com/grafana/grafana/pull/70693/files/080114d63ce67f5c49c6010c9270b2ce1a383dd5#diff-a24d5d308965e3b364ed7a21123ba8917b5f644d9940230e9de564bb82e63f76R83)

That sounds fine to me, any reason not to?

Reply:
I'm not a big fan of locking a mutex in one goroutine and unlocking in a different one, but it makes things cleaner I guess. I have changed it to use the mutex and removed the channel :+1: